File: July LeetCode Challenge/Day_9.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
	public int widthOfBinaryTree(TreeNode root) {
		if (root == null) {
			return 0;
		}
		Queue<TreeNode> queue = new LinkedList<>();
		int max = 1, leftmost = 1, rightmost = 1;
		root.val = 1;
		queue.offer(root);
		while (!queue.isEmpty()) {
			int size = queue.size();
			for (int i = 0; i < size; i++) {
				TreeNode cur = queue.poll();
				if (i == 0) {
					leftmost = cur.val;
				}
				if (i == size - 1) {
					rightmost = cur.val;
				}
				if (cur.left != null) {
					cur.left.val = cur.val * 2;
					queue.offer(cur.left);
				}
				if (cur.right != null) {
					cur.right.val = cur.val * 2 + 1;
					queue.offer(cur.right);
				}
			}
			max = Math.max(max, rightmost - leftmost + 1);

		}
		return max;

	}
}

File: October LeetCode Challenge/Day_9.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        return encode(root);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        stringstream ss(data);
        string item;
        TreeNode* root = NULL;

        while (getline (ss, item, '-'))
            root = insert(root, stoi(item));

        return root;
    }

private:
    string encode(TreeNode* root) {
        if (!root)
            return "";
        return to_string(root->val) + "-" + encode(root->left) + encode(root->right);
    }

    TreeNode* insert(TreeNode* root, int val) {
        if (root == NULL) {
            TreeNode* temp = new TreeNode(val);
            return temp;
        }

        if (val <= root->val)
            root->left = insert(root->left, val);

        else
            root->right = insert(root->right, val);

        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec* ser = new Codec();
// Codec* deser = new Codec();
// string tree = ser->serialize(root);
// TreeNode* ans = deser->deserialize(tree);
// return ans;